### PR TITLE
Splitting up DQN

### PR DESCRIPTION
**Pull request by sampreet-arthi**
Part of #169. Decided to split this into different PRs otherwise it might be take too long to merge. Also, there are some changes to the `BaseAgent` and `OnPolicyAgent` classes.

I've started refactoring the DQN class into separate files. The Vanilla DQN and Double DQN implementations train perfectly on basic envs right now.

### CODE REVIEW

File: genrl/deep/agents/base.py [hunk 1]
```diff
@@ -4,19 +4,26 @@
...
+        **kwargs
+    ):
         self.network_type = network_type
         self.env = env
-        self.epochs = epochs
+        self.batch_size = batch_size
+        self.layers = layers
+        self.gamma = gamma
         self.seed = kwargs.get("seed", None)
         self.render = kwargs.get("render", False)
```
Comment (Sharad24) [inline genrl/deep/agents/base.py:26]: No need to do a "get" operation, defeats the purpose of having ** in `__init__`

File: genrl/deep/agents/base.py [hunk 2]
```diff
@@ -82,22 +80,24 @@ def __init__(
...
-        epochs: int = 100,
         rollout_size: int = 2048,
         **kwargs
     ):
-        super(OnPolicyAgent, self).__init__(network_type, env, epochs, **kwargs)
-        self.batch_size = batch_size
-        self.gamma = gamma
+        super(OnPolicyAgent, self).__init__(
+            network_type,
+            env,
```
Comment (Sharad24) [inline genrl/deep/agents/base.py:92]: These two can be replaced with `*args`?

File: genrl/deep/agents/dqn/base.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+class DQN(BaseAgent):
+    def __init__(
+        self,
+        network_type: str,
+        env: Union[gym.Env, VecEnv],
+        batch_size: int = 32,
+        gamma: float = 0.99,
+        layers: Tuple = (32, 32),
+        lr: float = 0.001,
```
Comment (Sharad24) [inline genrl/deep/agents/dqn/base.py:29]: Should learning_rate be a part of the `BaseAgent`? Every agent needs it.

File: genrl/deep/agents/dqn/base.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    ) -> np.ndarray:
+
+        if not deterministic:
+            if np.random.rand() < self.epsilon:
+                return np.asarray(self.env.sample())
+
+        state = torch.FloatTensor(state)
+        q_values = self.model(state).detach().numpy()
+
+        return np.argmax(q_values, axis=-1)
```
Comment (Sharad24) [inline genrl/deep/agents/dqn/base.py:83]: For these type of functions it would be helpful to do a type check first, and cast them into the required type before computing. This would allow more options during policy updation, etc. 

For e.g. state could be a `np.ndarray` or a `torch.Tensor`?

File: genrl/deep/agents/dqn/base.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+            loss = F.mse_loss(q_values, target_q_values)
+            self.logs["value_loss"].append(loss.item())
+
+            self.optimizer.zero_grad()
+            loss.backward()
+            self.optimizer.step()
+
+            if timestep % update_interval == 0:
+                self.update_target_model()
```
Comment (Sharad24) [inline genrl/deep/agents/dqn/base.py:124]: These steps would be same in all DQNs? So we could put it in a different function altogether.
Would also reduce complexity

Comment (sampreet-arthi) [reply genrl/deep/agents/dqn/base.py:124]: Yeah. I changed this locally.

File: genrl/deep/agents/dqn/base.py (hunk 4 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/dqn/base.py:83]: There is a typecast being done here in `state = torch.FloatTensor(state)`

Comment (sampreet-arthi) [reply genrl/deep/agents/dqn/base.py:83]: If you mean before passing it to select_action I'll have to check after

File: genrl/deep/agents/dqn/base.py (hunk 3 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/dqn/base.py:29]: Yeah. Will change it when I make the Base Off Policy Agent. There would be an `lr_policy` and `lr_value` in the Base Agent.

File: genrl/deep/agents/base.py (hunk 1 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/base.py:26]: Ok, it was already there. I'll change it.

File: genrl/deep/agents/base.py (hunk 2 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/base.py:92]: I think that would get a little confusing and vague to users because the docs for that would be gone too. Are you suggesting we replace the init with `__init__(*args, **kwargs) `?

Comment (Sharad24) [reply genrl/deep/agents/base.py:92]: Docs are independent of this change? Will depend only on the docstring?

Comment (Sharad24) [reply genrl/deep/agents/base.py:92]: Yeah essentially all arguments from super's init can be managed with args, kwargs. So you can avoid network_type and env and anything else present in there

**Comment by sampreet-arthi** [comment]
I was actually thinking of this too. In the sense what are the general obvious components of any agent. And then we could break them down into functions accordingly.

**Comment by Sharad24** [comment]
> I was actually thinking of this too. In the sense what are the general obvious components of any agent. And then we could break them down into functions accordingly.

Great, awesome!

File: genrl/deep/agents/base.py (hunk 2 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/base.py:92]: Oh okay, I misunderstood. I'll do it soon.

File: genrl/deep/agents/base.py [hunk 6]
```diff
@@ -54,26 +52,26 @@ def select_action(self, state: np.ndarray) -> np.ndarray:
         :returns: Action based on the state and epsilon value
         :rtype: int, float, ...
         """
-        raise NotImplementedError()
-
-    def get_loss(self) -> None:
-        """
-        Calculate and return the total loss
-        """
-        raise NotImplementedError()
```
Comment (Het-Shah) [inline genrl/deep/agents/base.py:63]: Why is this removed?

File: genrl/deep/agents/dqn/base.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+    def create_model(self) -> None:
+        input_dim, action_dim, _, _ = get_env_properties(self.env, self.network_type)
+
+        self.model = get_model("v", self.network_type)(
+            input_dim, action_dim, "Qs", self.layers
+        )
+        self.target_model = deepcopy(self.model)
+
+        self.replay_buffer = PushReplayBuffer(self.replay_size)
```
Comment (Het-Shah) [inline genrl/deep/agents/dqn/base.py:62]: It would be good to take buffer type also as an input in `__init__`.

File: genrl/deep/agents/base.py (hunk 6 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/base.py:63]: The `OnPolicyAgent` uses `get_traj_loss` and the `OffPolicyAgent` will most likely use `get_q_loss` and `get_p_loss`. It's mostly that the name get_loss might be too general. I removed everything that's not being used right away. I'll add the functions back when needed.

File: genrl/deep/agents/dqn/base.py (hunk 7 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/dqn/base.py]: Ok, makes sense

File: genrl/deep/agents/dqn/base.py (hunk 4 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/dqn/base.py:88]: Doesn't seem like there's a problem here. `state` input is always `np.ndarray` and is always converted to `torch.FloatTensor` before being passed to the Q-function.

**Comment by sampreet-arthi** [comment]
I've replaced all the arguments in the DQN subclasses to use *args and **kwargs wherever possible. Should I do this to other important classes too? It might be a problem for writing documentation.

**Comment by Sharad24** [comment]
Which other subclasses? Also, why will it be a problem in writing documentation?

Also, take a look at code climate issues too, if possible. You'll know early-on what needs to be done. (14 new issues is a lot).

**Comment by sampreet-arthi** [comment]
Ok, a summary so far to make it clear. 

I've made one BaseDQN class. When you want to make a new DQN class, you can inherit from this and make changes to only the methods where necessary. There's issues with PrioritizedReplay and Categorical rn which I'll try to sort by tomorrow. The rest have been implemented and all of them solve CartPole-v0 in about 10k-20k timesteps. 

Certain Value functions like MlpDuelingValue, MlpNoisyValue and MlpCategoricalValue are added which will be used as the Q-functions. These will be available in dqn/values.py

Other snippets of code like the Double DQN target function and the helper functions are in dqn/utils.py

**Comment by sampreet-arthi** [comment]
These other classes (DoubleDQN, DuelingDQN etc) take arguments as 
```
__init__(*args, foo=bar,... , **kwargs) 
    super().__init__(*args, **kwargs) 
```

This looks compact for the DQN classes but I was suggesting adopting this for all other similar uses like in value and policy functions too. The problem with the docs would be that all the args wouldn't be in one place so it would be a pain to go around searching what all args are there. There would also be less transparency for users

**Comment by sampreet-arthi** [comment]
I'll check out the codeclimate issues soon. There's a lot of LGTM errors too.

**Comment by sampreet-arthi** [comment]
Let me know if this looks good. I'll get started on DDPG next.

**Comment by Sharad24** [comment]
Push a commit to relaunch checks

**Comment by sampreet-arthi** [comment]
Codeclimate errors are basically 6 errors (limit 4) and LGTM errors are `import torch` + `from torch import optim as opt`.

**Comment by sampreet-arthi** [comment]
This is ready to merge. @Het-Shah @Sharad24

File: genrl/deep/agents/a2c/a2c.py [hunk 8]
```diff
@@ -3,8 +3,8 @@
 import gym
 import numpy as np
 import torch
```
Comment (Sharad24) [inline genrl/deep/agents/a2c/a2c.py:5]: Not needed. `torch` will get imported with `from torch` statement.

File: genrl/deep/agents/dqn/base.py [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+import numpy as np
+import torch
+from torch import optim as opt
+from torch.nn import functional as F
+
+from genrl.deep.agents.base import OffPolicyAgent
+from genrl.deep.common.utils import get_env_properties, get_model, safe_mean
+
+
+class BaseDQN(OffPolicyAgent):
```
Comment (Sharad24) [inline genrl/deep/agents/dqn/base.py:13]: Why is this `BaseDQN`? Why not just `DQN`?

`BaseDQN` sounds a bit off, sounds like some ABC. I think you can change it to just `DQN` because all other are **extra** additions to the normal `DQN`

File: genrl/deep/agents/dqn/base.py [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+        """Function to calculate the loss of the Q-function
+
+        Returns:
+            loss (:obj:`torch.Tensor`): Calculateed loss of the Q-function
+        """
+        states, actions, rewards, next_states, dones = self.replay_buffer.sample(
+            self.batch_size
+        )
+        states = states.reshape(-1, *self.env.obs_shape)
+        actions = actions.reshape(-1, *self.env.action_shape).long()
```
Comment (Sharad24) [inline genrl/deep/agents/dqn/base.py:162]: If actions are discrete, won't this already be long? Is this type casting needed?

File: genrl/deep/agents/dqn/base.py [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+        Returns:
+            target_q_values (:obj:`torch.Tensor`): Target Q values for the DQN
+        """
+        next_q_target_values = self.target_model(next_states)
+        max_next_q_target_values = next_q_target_values.max(1)[0]
+
+        target_q_values = rewards + self.gamma * torch.mul(
+            max_next_q_target_values, (1 - dones)
+        )
+        return target_q_values.unsqueeze(-1)
```
Comment (Sharad24) [inline genrl/deep/agents/dqn/base.py:150]: Can you write down comments before squeeze and unsqueeze operations? 

The reason I'm suggesting this is because it will be easier to debug algorithms so I think its an important addition.

File: genrl/deep/agents/dqn/base.py [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+        Args:
+            state (:obj:`np.ndarray`): Current state of the environment
+            deterministic (bool): Should the policy be deterministic or stochastic
+
+        Returns:
+            action (:obj:`np.ndarray`): Action taken by the agent
+        """
+        if not deterministic:
+            if np.random.rand() < self.epsilon:
+                return np.asarray(self.env.sample())
```
Comment (Sharad24) [inline genrl/deep/agents/dqn/base.py:111]: `nit`
Returning inside if statements is not preferable generally as a programming practice.

File: genrl/deep/agents/dqn/base.py [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+class DQN(BaseDQN):
+    """Vanilla DQN class
+
+    Simply calls the `create_model` function from the base class
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(DQN, self).__init__(*args, **kwargs)
+        self.empty_logs()
+        self.create_model()
```
Comment (Sharad24) [inline genrl/deep/agents/dqn/base.py:246]: Same reason why I was suggesting to make the `BaseDQN` just `DQN` as this class does definition does not seem very useful.

File: genrl/deep/agents/dqn/categorical.py [hunk 14]
```diff
@@ -0,0 +1,144 @@
+from copy import deepcopy
+from typing import Tuple
+
+import numpy as np
+import torch
+from torch import optim as opt
```
Comment (Sharad24) [inline genrl/deep/agents/dqn/categorical.py:6]: `import torch` not required

File: genrl/deep/agents/dqn/categorical.py [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+            action (:obj:`np.ndarray`): Action taken by the agent
+        """
+        if not deterministic:
+            if np.random.rand() < self.epsilon:
+                return np.asarray(self.env.sample())
+
+        state = torch.FloatTensor(state)
+        dist = self.model(state).data.cpu()
+        dist = dist * torch.linspace(self.v_min, self.v_max, self.num_atoms)
+        action = dist.sum(2).max(1)[1].numpy()
```
Comment (Sharad24) [inline genrl/deep/agents/dqn/categorical.py:98]: Can you add a comment explaining this operation? Makes it more readable as there a bunch of operations. And it would help in debugging

File: genrl/deep/agents/dqn/categorical.py [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+        rewards = torch.FloatTensor(rewards).reshape(-1)
+        next_states = next_states.reshape(-1, *self.env.obs_shape)
+        dones = torch.FloatTensor(dones).reshape(-1)
+
+        projection_distribution = get_projection_distribution(
+            self, next_states, rewards, dones
+        )
+        dist = self.model(states)
+        actions = actions.unsqueeze(1).expand(-1, 1, self.num_atoms)
+        dist = dist.gather(1, actions).squeeze(1)
```
Comment (Sharad24) [inline genrl/deep/agents/dqn/categorical.py:121]: Comment before `actions = actions.unq`

File: genrl/deep/agents/dqn/double.py [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+            rewards (:obj:`list`): Rewards at each timestep for each environment
+            dones (:obj:`list`): Game over status for each environment
+
+        Returns:
+            target_q_values (:obj:`torch.Tensor`): Target Q values for the DQN
+        """
+        next_q_values = self.model(next_states)
+        next_best_actions = next_q_values.max(1)[1].unsqueeze(1)
+
+        rewards, dones = rewards.unsqueeze(-1), dones.unsqueeze(-1)
```
Comment (Sharad24) [inline genrl/deep/agents/dqn/double.py:54]: Needs comments. Improves readability for newbies.

File: genrl/deep/agents/dqn/prioritized.py [hunk 18]
```diff
@@ -0,0 +1,76 @@
+import torch
+from torch import optim as opt
+
+from genrl.deep.agents.dqn.base import BaseDQN
+
+
+class PrioritizedReplayDQN(BaseDQN):
```
Comment (Sharad24) [inline genrl/deep/agents/dqn/prioritized.py:7]: Why is this a separate class altogether? What if someone wants to use `CategoricalDQN` with `PrioritizedReplayBuffer`?

File: genrl/deep/agents/dqn/values.py [hunk 19]
```diff
@@ -0,0 +1,237 @@
...
+
+    def forward(self, state: torch.Tensor) -> torch.Tensor:
+        features = self.feature(state)
+        advantage = self.advantage(features)
+        value = self.value(features)
+        return value + advantage - advantage.mean()
+
+
+class CnnDuelingValue(nn.Module):
+    """Class for Dueling DQN's MLP Q-Value function
```
Comment (Sharad24) [inline genrl/deep/agents/dqn/values.py:35]: Both can be generalised and moved to `values.py`?

File: genrl/deep/agents/dqn/values.py [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+    def forward(self, state: torch.Tensor) -> torch.Tensor:
+        return self.model(state)
+
+    def reset_noise(self) -> None:
+        for layer in self.model:
+            if isinstance(layer, NoisyLinear):
+                layer.reset_noise()
+
+
+class CnnNoisyValue(nn.Module):
```
Comment (Sharad24) [inline genrl/deep/agents/dqn/values.py:91]: These two as well?

File: genrl/deep/agents/dqn/values.py [hunk 21]
```diff
@@ -0,0 +1,237 @@
...
+        inp = self.model(inp)
+        return inp
+
+    def reset_noise(self) -> None:
+        for layer in self.model:
+            if isinstance(layer, NoisyLinear):
+                layer.reset_noise()
+
+
+class MlpCategoricalValue(nn.Module):
```
Comment (Sharad24) [inline genrl/deep/agents/dqn/values.py:128]: Ok, all of these? What do you think?

File: genrl/deep/agents/dqn/values.py [hunk 22]
```diff
@@ -0,0 +1,237 @@
...
+    """
+    Returns DQN model given the type of DQN
+
+    :param name_: Name of the policy needed
+    :type name_: str
+    :returns: DQN class to be used
+    """
+    if name_ in dqn_value_registry:
+        return dqn_value_registry[name_]
+    raise NotImplementedError
```
Comment (Sharad24) [inline genrl/deep/agents/dqn/values.py:237]: I definitely like this, but I'd prefer these to be part of general value function definitions since at the end of the day, they can be used right?

File: genrl/deep/agents/ppo1/ppo1.py [hunk 23]
```diff
@@ -3,8 +3,8 @@
 import gym
 import numpy as np
 import torch
-import torch.nn as nn
-import torch.optim as opt
+from torch import nn as nn
```
Comment (Sharad24) [inline genrl/deep/agents/ppo1/ppo1.py:6]: `import torch` not required

**Comment by Sharad24** [comment]
Looks great on the code climate end, so need to worry about them for now

File: genrl/deep/agents/a2c/a2c.py (hunk 8 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/a2c/a2c.py:5]: Ok I didn't know that 😬

File: genrl/deep/agents/dqn/base.py (hunk 10 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/dqn/base.py:162]: Yeah, when we get it from the buffer, it's converted to a FloatTensor by default. The `.gather` function needs it to be a LongTensor, hence the typecast.

File: genrl/deep/agents/dqn/base.py (hunk 11 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/dqn/base.py:150]: Ok I'll do that. In fact, I thought this was one of the main problems with the DQN. Random squeezing, unsqueezing and reshaping everywhere.

File: genrl/deep/agents/dqn/base.py (hunk 12 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/dqn/base.py:111]: Ok, will change this

File: genrl/deep/agents/dqn/categorical.py (hunk 15 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/dqn/categorical.py:98]: Yeah ok

File: genrl/deep/agents/dqn/prioritized.py (hunk 18 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/dqn/prioritized.py:7]: Yeah, so what I was actually thinking of was these are all model classes.

If someone wants to a combination of these, then they should be able to make their own custom DQN by adding in custom snippets of code. Like for example the `ddqn_q_target` or the `projection_distribution`. I can segment the code into even more functions so that it gets easier to use. I'll put an example up today. 

For specifically PrioritizedBuffer, i was thinking we can make that available for all OffPolicy algos to use (idk about SAC but DDPG and TD3 had slight variants allowing them to do so). It's not standard but it's an additional feature.

File: genrl/deep/agents/dqn/values.py (hunk 22 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/dqn/values.py:237]: It's just that there are 6 different functions. The values.py file would become really big. Also, can we use them anywhere else? (Except in DQN) 

Also, I was also thinking that this seemed redundant to develop so many different classes and was thinking of simplifying this too into snippets. (Otherwise how do you use MlpDuelingValue and MlpNoisyValue together?) It would be really messy if we had to implement Rainbow and had to make a MlpDuelingNoisyCategoricalValue :p

Comment (Sharad24) [reply genrl/deep/agents/dqn/values.py:237]: I'm not sure about Dueling but noisy networks are pretty common to enhance exploration. For e.g. Categorical can be used in Distributional DDPG?

File: genrl/deep/agents/dqn/base.py (hunk 10 above)
Comment (Sharad24) [reply genrl/deep/agents/dqn/base.py:162]: Ok

File: genrl/deep/agents/a2c/a2c.py (hunk 8 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/a2c/a2c.py:5]: This isn't happening. I still get an error saying `torch` isn't defined.

File: genrl/deep/agents/dqn/base.py (hunk 9 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/dqn/base.py:13]: Yeah, changed this to `DQN`. You were right, the extra class is redundant.

File: genrl/deep/agents/dqn/values.py (hunk 22 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/dqn/values.py:237]: Ok, I'll move these if `values.py` doesn't get too big, otherwise I'll see what else we can do.

File: genrl/deep/agents/a2c/a2c.py (hunk 8 above)
Comment (sampreet-arthi) [reply genrl/deep/agents/a2c/a2c.py:5]: EDIT: It does import torch but not `torch.Tensor` (used for typing hints). So either we just import torch or we use `from torch import Tensor` and use just `Tensor` which can be misunderstood. What do you think?

**Comment by sampreet-arthi** [comment]
I've added a `create_model` option like @Sharad24 said. See if this looks fine.

**Comment by sampreet-arthi** [comment]
Some things left to do:
- [ ] Add comments to explain operations
- [ ] Get rid of `torch` import errors properly
- [ ] Compartmentalize code to facilitate easy usage with OffPolicyTrainer later on
- Categorical DQN not training. (Will take care of this in a separate issue)

**Comment by sampreet-arthi** [comment]
This maybe minor but for the torch errors, there are 3 alternatives:
- Using only `import torch` and the rest will have to be used as `torch.optim.Adam`
    - Long names which will be confusing and look dirty
- Using only `from torch import ... as ...` which will mean that `torch.Tensor` will have to be replaced as `Tensor` and we'll have to replace every usage of torch like `torch.clamp` etc with the `from import` statement
    - Obviously cumbersome and will quickly lead to a lot of bugs if someone else starts editing the files
- Switch off the option by making LGTM ignore
    - Best solution as it doesn't seem like a big problem to worry about

**Comment by Sharad24** [comment]
Ok lets do the LGTM ignore then?

**Comment by sampreet-arthi** [comment]
Yeah. Also, do check the code once again, there were a lot of changes mainly in common/values.py and dqn/utils.py

**Comment by sampreet-arthi** [comment]
I'm starting on DDPG. Let me know if this looks good. @Het-Shah @Sharad24

**Comment by sampreet-arthi** [comment]
I don't think I'd changed any dependencies, could you take a look at the GH-Actions build error?

**Comment by sampreet-arthi** [comment]
Ready to merge @Sharad24 

Solves #150. #146 doesn't seem to be a problem.
Will separate issues to tackle the following (as they would take time or would need to edit non-DQN files):
- Ensure training of Categorical DQN
- Add CUDA support (will be done for all Off Policy Agents at once)
- Check training for Atari envs
- Using Dueling + Noisy or Prioritized + Categorical etc needs to be clarified. (will be done after Off Policy agents' structures are more sorted)
- Add support for Prioritized Experience Replay for all OffPolicy algos (not a priority atm)

**Comment by Sharad24** [comment]
Awesome :)

Also, sorry for being so picky about this.

**Comment by Sharad24** [comment]
Please merge if you don't want to add anything else

**Comment by sampreet-arthi** [comment]
Yep, finally done :p
